selectable_pdf_infra/lambda/selectable_pdf/main.py

```diff
@@ -0,0 +1,356 @@
+1. load the original PDF from S3 (S3 localtion info located in the SQS message)
+2. read all the page and convert them to image. Indeed, the original PDF might already 
+   have selectable characters, or a mix with pixel characters, so we want to avoid 
+   overlay characters on characters
+3. Add the character to the images. the characters are added by words. For each word, 
+   the best fitting fontsize is computed to the character word length equal the pixel 
+   word length
+4. Save the pdf with selectable characters
+
+Required environment variable
+* OUTPUT_BUCKET
+* DDB_DOCUMENTS_TABLE
+'''
+
+# import modules
+# --------------
+# standard modules
+import os
+import boto3
+import json
+import logging
+import datetime
+import math
+
+# custom modules from layers
+import fitz  #This is PyMuPdf
+from textracttools import load_json_from_s3
+
+# typing
+from typing import Dict, Optional, List, Any
+
+
+# logger
+# ------ 
+#If no LOG_LEVEL env var or wrong LOG_LEVEL env var, fallback 
+# to INFO log level
+log_level = os.getenv('LOG_LEVEL', default='INFO')
+log_level_int = int()
+if log_level=='WARNING':
+    log_level_int = logging.WARNING
+elif log_level=='ERROR':
+    log_level_int = logging.WARNING
+elif log_level=='DEBUG':
+    log_level_int = logging.DEBUG
+else:
+    log_level_int = logging.INFO
+logging.basicConfig(
+    format='%(levelname)s %(message)s',
+    level=log_level_int,
+    force=True  #new in py3.8
+)
+logger = logging.getLogger()
+
+
+# lambda entrypoint
+# -----------------
+def lambda_handler(event, context):
+    logger.info('event: {}'.format(event))
+    args = parse_args(event)
+    logger.info(f'args: {args}')
+
+    # Get AWS connectors.
+    ddb_ress = boto3.resource('dynamodb')
+    ddb_doc_table = ddb_ress.Table(args['ddb_documents_table'])
+
+    returns = list()
+    for rec in args['records']:
+        document_id = rec['document_id']
+        document_name = rec['document_name']
+        logger.info('document id: {}'.format(document_id))
+        logger.info('document name: {}'.format(document_name))
+        logger.info(f"document bucket: {rec['original_document_s3']['bucket']}")
+        logger.info(f"document key: {rec['original_document_s3']['key']}")
+
+        # store info about starting creating the selectable PDF document
+        try:
+            ddb_doc_table.update_item(
+                Key={
+                    'document_id': document_id,  # HASH key
+                    'document_name': document_name  # RANGE key
+                },
+                UpdateExpression='SET selectable_pdf=:att1',  # This will set a new attribute
+                ExpressionAttributeValues={
+                    ':att1': {
+                        'datetime': datetime.datetime.now().strftime('%Y-%m-%dT%H:%M:%S') + '+00:00'
+                    }
+                }
+            )
+        except Exception as ex:
+            logger.error(f'Cannot update item in DynamoDB table {args["ddb_documents_table"]}')
```

Comment:
Python convention is to use double quotes everywhere, I suggest to be consistent at least

Reply:
Python PEP8 does not enforce single or double quote for strings (see "string quote" here https://www.python.org/dev/peps/pep-0008/#string-quotes). I use single quote in Python. Nevertheless, the f-string formatting is leading to a few issues. Normally you use f-string formatting as follow 
```
zero = 0
my_str = f'this value is {zero}'
```
But if the insertion is a dict key, then there are three options:
```
mydict['zero'] = 0
my_str1 = f'this value is {mydict["zero"]}'  #option 1: double quote for the key
my_str2 = f"this value is {mydict['zero']}"  #option 2: single quote for the key
my_str3 = f'this value is {mydict[\'zero\']}' #option 3: single quote everywhere, but escapes char required around the key
```
Or you can use a 4th option, the "old" formatting:
```
mydict['zero'] = 0
my_str1 = 'this value is {val}'.format(val=mydict['zero']) #option 4
```
I think option 1 and 2 are the most readable. As they both mix quote type, I randomly choose option 2. I update the code with option 2.

Reply:
Ok, sounds reasonable